common/src/main/java/io/netty/util/concurrent/Futures.java

```diff
@@ -0,0 +1,208 @@
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    R result = (R) completed.getNow();
+                    recipient.trySuccess(result);
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class CallableMapper<R, T> implements Callable<R> {
+        private final Future<T> future;
+        private final Function<T, R> mapper;
+
+        CallableMapper(Future<T> future, Function<T, R> mapper) {
+            this.future = future;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public R call() throws Exception {
+            return mapper.apply(future.getNow());
+        }
+    }
+
+    private static class Mapper<R, T> implements FutureListener<Object> {
+        private final Promise<R> recipient;
+        private final Function<T, R> mapper;
+
+        Mapper(Promise<R> recipient, Function<T, R> mapper) {
+            this.recipient = recipient;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void operationComplete(Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    T result = (T) completed.getNow();
+                    R mapped = mapper.apply(result);
+                    recipient.trySuccess(mapped);
+                } catch (RuntimeException e) {
+                    tryFailure(recipient, e, LOGGER);
+                }
+            } else {
+                propagateUncommonCompletion(completed, recipient);
+            }
+        }
+    }
+
+    private static class FlatMapper<R, T> implements FutureListener<Object> {
+        private final Promise<R> recipient;
+        private final Function<T, Future<R>> mapper;
+
+        FlatMapper(Promise<R> recipient, Function<T, Future<R>> mapper) {
+            this.recipient = recipient;
+            this.mapper = mapper;
+        }
+
+        @Override
+        public void operationComplete(Future<?> completed) throws Exception {
+            if (completed.isSuccess()) {
+                try {
+                    @SuppressWarnings("unchecked")
+                    T result = (T) completed.getNow();
+                    Future<R> future = mapper.apply(result);
+                    if (future.isSuccess()) {
+                        recipient.trySuccess(future.getNow());
+                    } else {
+                        future.addListener(recipient, passThrough());
+                        recipient.addListener(future, PromiseNotifier::propagateCancel);
+                    }
+                } catch (Exception e) {
```

Comment:
Let's use `Throwable` as well here (see above)
```suggestion
                } catch (Throwable e) {
```